onnxruntime/core/providers/cuda/tensor/gather_nd.cc

```diff
@@ -38,16 +38,31 @@ Status CheckBatchDimensionsMatch(
   return Status::OK();
 }
 
-#define TYPED_FUNCTION_CALL_FWD(T)                                                                                                            \
-  if (T_type == DataTypeImpl::GetType<T>()) {                                                                                                 \
-    GatherNDImpl<ToCudaType<T>::MappedType>(num_slices, kernel_input_data, kernel_output_data, slice_size, input_slice_offsets_buffer.get()); \
-  }
-
-#define TYPED_FUNCTION_CALL_BWD(T)                                                                                                                \
-  if (T_type == DataTypeImpl::GetType<T>()) {                                                                                                     \
-    GatherNDGradImpl<ToCudaType<T>::MappedType>(num_slices, kernel_input_data, kernel_output_data, slice_size, input_slice_offsets_buffer.get()); \
+template <typename T>
+struct ComputeImpl {
+  void operator()(const int64_t num_slices,
+                  const int64_t slice_size,
+                  const void* const kernel_input_data,
+                  const bool fwd,
+                  void* const kernel_output_data,
+                  int64_t* const input_slice_offsets_data) const {
+    typedef typename ToCudaType<T>::MappedType CudaT;
+    if (fwd) {
+      GatherNDImpl<CudaT>(num_slices, kernel_input_data,
+                          kernel_output_data, slice_size,
+                          input_slice_offsets_data);
+    } else {
+#ifdef ENABLE_TRAINING
+      GatherNDGradImpl<CudaT>(num_slices, kernel_input_data,
```

Comment:
The refactoring I was thinking was to avoid exposing this training related function in the inference code. 
Can we make GatherNDBase::CommonComputeKernel() accepts a functor, so that fw code can pass GatherNDImpl and bw code can pass GatherNDGradImpl?
This way, we don't need to do the if(fwd) switch

Reply:
I just realized this, yes I will do it in the new PR

Reply:
To do this, I would prefer to do another way. Want to change GatherNDBase::CommonComputeKernel()  to GatherNDBase::PrepareCompute(), and put kernel invocation part out of it. Let each kernels (CUDA GatherND, GATHERNDGrad, CPU GatherND, GatherNDGrad) to share this part, each kernel will have the invocation part separately.

Reply:
Sure. That's a even better idea